### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry): residue field of a point and evaluation

### PR DESCRIPTION
**Pull request by chrisflav**
Defines the residue field of a point of a locally ringed space and the corresponding evaluation map.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/ResidueField.lean [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  mem_basicOpen_iff_evaluation_ne_zero X f ⟨x, trivial⟩
+
+variable {X Y : LocallyRingedSpace.{u}} (f : X ⟶ Y)
+
+/-- If `X ⟶ Y` is a morphism of locally ringed spaces and `x` a point of `X`, we obtain
+a morphism of residue fields in the other direction. -/
+def evaluationMap (x : X) : Y.residueField (f.val.base x) ⟶ X.residueField x :=
+  LocalRing.ResidueField.map (LocallyRingedSpace.stalkMap f x)
+
+lemma evaluation_naturality {V : Opens Y} (x : f ⁻¹ᵁ V) :
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/ResidueField.lean:95]:
```suggestion
@[reassoc]
lemma evaluation_naturality {V : Opens Y} (x : f ⁻¹ᵁ V) :
```

File: Mathlib/RingTheory/Ideal/LocalRing.lean [hunk 2]
```diff
@@ -369,6 +369,19 @@ def residue : R →+* ResidueField R :=
   Ideal.Quotient.mk _
 #align local_ring.residue LocalRing.residue
 
+variable {R}
+
+lemma residue_ker : RingHom.ker (residue R) = maximalIdeal R :=
```
Comment (erdOne) [inline Mathlib/RingTheory/Ideal/LocalRing.lean:374]:
```suggestion
lemma ker_residue : RingHom.ker (residue R) = maximalIdeal R :=
```

File: Mathlib/RingTheory/Ideal/LocalRing.lean [hunk 3]
```diff
@@ -369,6 +369,19 @@ def residue : R →+* ResidueField R :=
...
+  Ideal.mk_ker
+
+lemma residue_eq_zero_iff (x : R) : residue R x = 0 ↔ x ∈ maximalIdeal R := by
+  rw [← RingHom.mem_ker, residue_ker]
+
+lemma residue_ne_zero_iff_isUnit (x : R) : residue R x ≠ 0 ↔ IsUnit x := by
+  rw [← not_iff_not, ne_eq, not_not, residue_eq_zero_iff, mem_maximalIdeal, mem_nonunits_iff]
```
Comment (erdOne) [inline Mathlib/RingTheory/Ideal/LocalRing.lean:381]: Do you think these should be simp lemmas?

File: Mathlib/AlgebraicGeometry/ResidueField.lean [hunk 4]
```diff
@@ -0,0 +1,119 @@
+/-
+Copyright (c) 2024 Christian Merten. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christian Merten
+-/
+import Mathlib.AlgebraicGeometry.OpenImmersion
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/ResidueField.lean:6]: Why is this necessary? I think everything `LocallyRingedSpace`-related should be in `Geometry/RingedSpace/OpenImmersion`.

File: Mathlib/AlgebraicGeometry/ResidueField.lean [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+import Mathlib.AlgebraicGeometry.GammaSpecAdjunction
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/ResidueField.lean:7]: We should split this file and move the `LocallyRingedSpace` adjunction away... (not in this PR)

File: Mathlib/AlgebraicGeometry/GammaSpecAdjunction.lean [hunk 6]
```diff
@@ -69,6 +69,18 @@ def ΓToStalk (x : X) : Γ.obj (op X) ⟶ X.presheaf.stalk x :=
   X.presheaf.germ (⟨x, trivial⟩ : (⊤ : Opens X))
 #align algebraic_geometry.LocallyRingedSpace.Γ_to_stalk AlgebraicGeometry.LocallyRingedSpace.ΓToStalk
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/GammaSpecAdjunction.lean:70]: Or the real problem is that this def doesn't belong here.

**Review by erdOne — COMMENTED** [review]
Please move the file to `Geometry/RingedSpace/LocallyRingedSpace`.

File: Mathlib/RingTheory/Ideal/LocalRing.lean (hunk 3 above)
Comment (chrisflav) [reply Mathlib/RingTheory/Ideal/LocalRing.lean:381]: Probably makes sense.

**Comment by chrisflav** [comment]
Thanks for the quick review @erdOne! An unfortunate side effect of reducing the imports, is that the nice notations you introduced are no longer available. I think we should move these down the import hierarchy.

File: Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    X.Γevaluation x f = 0 ↔
+      (X.presheaf.germ ⟨x, show x ∈ ⊤ by trivial⟩) f ∈ LocalRing.maximalIdeal (X.stalk x) :=
+  LocalRing.residue_eq_zero_iff _
+
+lemma Γevaluation_ne_zero_iff_isUnit (x : U) (f : X.presheaf.obj (op U)) :
+    X.evaluation x f ≠ 0 ↔ IsUnit ((X.presheaf.germ x) f) :=
+  LocalRing.residue_ne_zero_iff_isUnit _
+
+@[simp]
+lemma evaluation_ne_zero_iff_mem_basicOpen (f : X.presheaf.obj (op U)) (x : U) :
```
Comment (erdOne) [inline Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean:79]: maybe `evaluation_eq_zero_iff_not_mem_basicOpen` as simp instead.

Comment (chrisflav) [reply Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean:79]: Instead or additionally?

Comment (erdOne) [reply Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean:79]: I thought lint was complaining that the LHS is not simpNF as `ne_eq` rewrites it to `\not _ = _`

Comment (chrisflav) [reply Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean:79]: No, it's only complaining for the `Γ` version, because it rewrites with `Γ_obj`.

Comment (erdOne) [reply Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean:79]: Ah okay. `evaluation_eq_zero_iff_not_mem_basicOpen` can still be added as simp lemma though. And then this lemma would be provable by simp.

Comment (chrisflav) [reply Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean:79]: Should I still keep the `evaluation_ne_zero_iff_mem_basicOpen` version?

Comment (erdOne) [reply Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean]: I kind of believe that `evaluation_eq_zero_iff_not_mem_basicOpen` and `evaluation_ne_zero_iff_mem_basicOpen` (and their `Γ`-counterparts) are the only `evaluation_*_zero_iff_*` lemmas worth keeping here, since `mem_basicOpen` should currently be our simpNF of this concept.

Comment (erdOne) [reply Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean]: Not 100% sure though. Will need to use this API more to find out, but this is my gut feeling.

Comment (chrisflav) [reply Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean]: I _thought_ I did use the `mem_maximalIdeal` versions somewhere in my application, but turns out I did not. So I am happy to remove them.

Comment (erdOne) [reply Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean]: Even if you did, I believe you should do that in two rewrites (or `A.trans B.symm`). It is more manageable for both mathlib and users if there is one canonical way to spell everything. This leads to n rewrite lemmas for n equivalent spellings instead of n^2, and the user will then always be encouraged to change their arguments into the canonical form, reliably knowing that this where all the API is.

Comment (chrisflav) [reply Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean]: Understood.

File: Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+If we interpret sections over `U` as functions of `X` defined on `U`, then this ring map
+corresponds to evaluation at `x`.
+-/
+def evaluation (x : U) :
+    X.presheaf.obj (op U) ⟶ X.residueField x :=
+  X.presheaf.germ x ≫ LocalRing.residue _
+
+/-- The global evaluation map from `Γ(X, ⊤)` to the residue field at `x`. -/
+def Γevaluation (x : X) : LocallyRingedSpace.Γ.obj (op X) ⟶ X.residueField x :=
```
Comment (erdOne) [inline Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean:58]:
```suggestion
def Γevaluation (x : X) : X.presheaf.obj (op ⊤) ⟶ X.residueField x :=
```
Does this make `Γevaluation_eq_zero_iff_not_mem_basicOpen` a valid simp lemma?

Comment (chrisflav) [reply Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean:58]: Indeed, good catch. Changed.

File: Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+corresponds to evaluation at `x`.
+-/
+def evaluation (x : U) : X.presheaf.obj (op U) ⟶ X.residueField x :=
+  X.presheaf.germ x ≫ LocalRing.residue _
+
+/-- The global evaluation map from `Γ(X, ⊤)` to the residue field at `x`. -/
+def Γevaluation (x : X) : X.presheaf.obj (op ⊤) ⟶ X.residueField x :=
+  X.evaluation ⟨x, show x ∈ ⊤ from trivial⟩
+
+@[simp]
```
Comment (joelriou) [inline Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean:60]: Do we really want this to be a simp lemma? (same for `Γevaluation_eq_zero_iff_not_mem_basicOpen`.)

Comment (erdOne) [reply Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean:60]: I think we should choose a simpNF. Do you think `X.evaluation x f = 0` would be better?

File: Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+lemma Γevaluation_ne_zero_iff_mem_basicOpen (x : X) (f : X.presheaf.obj (op ⊤)) :
+    X.Γevaluation x f ≠ 0 ↔ x ∈ X.toRingedSpace.basicOpen f :=
+  evaluation_ne_zero_iff_mem_basicOpen X ⟨x, show x ∈ ⊤ by trivial⟩ f
+
+variable {X Y : LocallyRingedSpace.{u}} (f : X ⟶ Y)
+
+/-- If `X ⟶ Y` is a morphism of locally ringed spaces and `x` a point of `X`, we obtain
+a morphism of residue fields in the other direction. -/
+def evaluationMap (x : X) : Y.residueField (f.val.base x) ⟶ X.residueField x :=
```
Comment (joelriou) [inline Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean:83]: Could you also include a lemma which relates `stalkMap f x` and `evaluationMap f x`, and lemmas `evaluationMap_id` and `evaluationMap_comp`?

Comment (joelriou) [reply Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean:83]: (Also, I am not sure about the name "evaluationMap"; do not you feel that `mapResidueField` would be better?)

Comment (chrisflav) [reply Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean:83]: `residueFieldMap` (in accordance with `stalkMap`) maybe?

File: Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean (hunk 9 above)
Comment (joelriou) [reply Mathlib/Geometry/RingedSpace/LocallyRingedSpace/ResidueField.lean:60]: Ok, let us keep these simp lemmas.

**Comment by erdOne** [comment]
Thanks!
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Comment by chrisflav** [comment]
Thanks for the reviews!
